File: javascript/rational-numbers/rational-numbers.js
export class Rational {
    constructor(a, b) {
        this.numerator = a;
        this.denominator = b;

        this.reduce();
    }

    add(other) {
        const a = this.numerator * other.denominator + other.numerator * this.denominator;
        const b = this.denominator * other.denominator;

        return new Rational(a, b);
    }

    sub(other) {
        const a = this.numerator * other.denominator - other.numerator * this.denominator;
        const b = this.denominator * other.denominator;

        return new Rational(a, b);
    }

    mul(other) {
        const a = this.numerator * other.numerator;
        const b = this.denominator * other.denominator;

        return new Rational(a, b);
    }

    div(other) {
        const a = this.numerator * other.denominator;
        const b = this.denominator * other.numerator;

        return new Rational(a, b);
    }

    abs() {
        return new Rational(Math.abs(this.numerator), this.denominator);
    }

    exprational(n) {
        return new Rational(this.numerator ** n, this.denominator ** n);
    }

    expreal(n) {
        const result = n ** (this.numerator / this.denominator);

        return result.round(2);
    }

    reduce() {
        const divisor = gcd(this.numerator, this.denominator);

        this.numerator /= divisor;
        this.denominator /= divisor;

        if (this.denominator < 0) {
            this.numerator = -this.numerator;
            this.denominator = -this.denominator;
        }

        return this;
    }
}

function gcd(n, m) {
    if (m === 0)
        return n;
    else
        return gcd(m, n % m);
}

// Source: <https://stackoverflow.com/a/41716722/11588119>
Number.prototype.round = function(n = 0) {
    const scale = 10 ** n;

    return Math.round((this + Number.EPSILON) * scale) / scale;
}

File: csharp/rational-numbers/RationalNumbers.cs
using System;
using System.Diagnostics;

public static class Helpers
{
    public static int IntPow(int a, int b)
    {
        return (int)Math.Pow(a, b);
    }

    public static int GCD(int a, int b)
    {
        if (b == 0)
            return a;
        else
            return GCD(b, a % b);
    }
}

public static class RealNumberExtension
{
    public static double Expreal(this int realNumber, RationalNumber r)
    {
        return Math.Exp(Math.Log(realNumber) * r.Numerator / r.Denominator);
    }
}

public struct RationalNumber
{
    public int Numerator, Denominator;

    public RationalNumber(int numerator, int denominator)
    {
        Numerator = numerator;
        Denominator = denominator;

        Reduce();
    }

    public static RationalNumber operator +(RationalNumber r1, RationalNumber r2)
    {
        var a = r1.Numerator * r2.Denominator + r2.Numerator * r1.Denominator;
        var b = r1.Denominator * r2.Denominator;

        return new RationalNumber(a, b);
    }

    public static RationalNumber operator -(RationalNumber r1, RationalNumber r2)
    {
        var a = r1.Numerator * r2.Denominator - r2.Numerator * r1.Denominator;
        var b = r1.Denominator * r2.Denominator;

        return new RationalNumber(a, b);
    }

    public static RationalNumber operator *(RationalNumber r1, RationalNumber r2)
    {
        var a = r1.Numerator * r2.Numerator;
        var b = r1.Denominator * r2.Denominator;

        return new RationalNumber(a, b);
    }

    public static RationalNumber operator /(RationalNumber r1, RationalNumber r2)
    {
        var a = r1.Numerator * r2.Denominator;
        var b = r1.Denominator * r2.Numerator;

        return new RationalNumber(a, b);
    }

    public RationalNumber Abs()
    {
        return new RationalNumber(Math.Abs(Numerator), Denominator);
    }

    public RationalNumber Reduce()
    {
        var divisor = Helpers.GCD(Numerator, Denominator);

        Numerator /= divisor;
        Denominator /= divisor;

        if (Denominator < 0) {
            Numerator = -Numerator;
            Denominator = -Denominator;
        }

        return this;
    }

    public RationalNumber Exprational(int power)
    {
        return new RationalNumber(Helpers.IntPow(Numerator, power),
                                  Helpers.IntPow(Denominator, power));
    }

    public double Expreal(int baseNumber)
    {
        return Math.Pow(baseNumber, Numerator / Denominator);
    }
}
